### PR TITLE
SS artifact and upgrading samtools/bcftools/htslib to 1.9

### PR DESCRIPTION
**Pull request by NagaComBio**
Major points,
- Upgrading to BCFtools/SAMtools/htslib to 1.9
- Calling variants using BCFtools mpileup and samtools mpileup for lookup in control (similar to the old way)
- PE overlaps are removed as before
- Removing supplementary reads and left-out duplicate reads from variant calling
- Deactivating the use to ExAC for no-control workflow filtering.
- Updating to new local control and change the max AF threshold to 0.01

### CODE REVIEW

File: resources/analysisTools/snvPipeline/filter_PEoverlap.py [hunk 1]
```diff
@@ -80,19 +80,30 @@ def getIndexACGTNacgtn(is_reverse, is_read1, base):
...
         else:
             if DP4af > 0: DP4af -= 1   
     return(DP4rf, DP4rr, DP4af, DP4ar)
 
+def decrease_counter(count_list, remove_base, ref):
+	"""
+	Helper function in counting the criteria used to decrease DP4
+	"""
+	if remove_base == ref:
+		count_list[0] += 1
```
[the remark this answers is not in the document]
Comment (NagaComBio) [reply resources/analysisTools/snvPipeline/filter_PEoverlap.py]: That's a good idea, I will update this here and for rest of the script

File: resources/analysisTools/snvPipeline/filter_PEoverlap.py [hunk 2]
```diff
@@ -261,13 +264,14 @@ def performAnalysis(args):
...
                                         else:
                                             ACGTNacgtn2[ACGTNacgtn_index[1]] += 1
-
-                                        if(pileupread.alignment.is_supplementary):
-                                            remove_base = pileupread.alignment.seq[pileupread.query_position]	
-                                            remove_is_reverse = pileupread.alignment.is_reverse                                            
-                                            count_supple = decrease_counter(count_supple, remove_base, REF)                                            
-                                            (DP4rf, DP4rr, DP4af, DP4ar) = decreaseDP4(remove_base, remove_is_reverse, REF, ALT, DP4rf, DP4rr, DP4af, DP4ar)
-                                            continue
+
```
[the remark this answers is not in the document]
Comment (NagaComBio) [reply resources/analysisTools/snvPipeline/filter_PEoverlap.py]: This is comment is for why it is commented out, will update the comment with bit more clarity.

File: resources/analysisTools/snvPipeline/filter_PEoverlap.py [hunk 3]
```diff
@@ -80,19 +80,33 @@ def getIndexACGTNacgtn(is_reverse, is_read1, base):
...
             if DP4rf > 0: DP4rf -= 1
-    elif remove_base == ALT:
+    elif remove_base.lower() == ALT.lower():
         if remove_is_reverse:
             if DP4ar > 0: DP4ar -= 1
         else:
             if DP4af > 0: DP4af -= 1   
     return(DP4rf, DP4rr, DP4af, DP4ar)
 
+def decrease_counter(count_list, remove_base, ref):
```
[the remark this answers is not in the document]
Comment (NagaComBio) [reply resources/analysisTools/snvPipeline/filter_PEoverlap.py]: That's a neat idea, never implemented python-class in production, I would go for something like this

```
class BoolCounter:
  def __init__(self):
    self.ref = 0
    self.alt = 0

  def update(self, remove):
    if (remove):
      self.ref += 1
    else:
      self.alt += 1

  def counted(self):
    return [self.ref, self.alt]
```

File: resources/analysisTools/snvPipeline/filter_PEoverlap.py [hunk 4]
```diff
@@ -393,15 +394,15 @@ def performAnalysis(args):
...
                         remove_is_reverse = decreaseInfo[1]
 
-                        count_duplicate = decrease_counter(count_duplicate, remove_base, REF)
+                        count_duplicate.update(remove_base == REF)
                         (DP4rf, DP4rr, DP4af, DP4ar) = decreaseDP4(remove_base, remove_is_reverse, REF, ALT, DP4rf, DP4rr, DP4af, DP4ar)
 
             nonREFnonALT = nonREFnonALTrev + nonREFnonALTfwd
             #Format: DP2 -> "reference(forward + reverse), alt(forward + reverse)"
-            supple_dup_str = 'DP2sup=' + ','.join(map(str, count_supple))
-            supple_dup_str += ';DP2dup=' + ','.join(map(str, count_duplicate))
```
Comment (GWarsow) [inline resources/analysisTools/snvPipeline/filter_PEoverlap.py]: why did you remove the semicolon in this string append?

Comment (NagaComBio) [reply resources/analysisTools/snvPipeline/filter_PEoverlap.py]: It looks like a bug, it needs semicolon here.

File: resources/analysisTools/snvPipeline/environments/tbi-lsf-cluster.sh [hunk 5]
```diff
@@ -12,12 +12,15 @@ set -ue
...
 module load samtools/"${SAMTOOLS_VERSION:?SAMTOOLS_VERSION undefined}"
+module load bcftools/"${BCFTOOLS_VERSION:?BCFTOOLS_VERSION undefined}"
 module load htslib/"${HTSLIB_VERSION:?HTSLIB_VERSION undefined}"
 module load perl/"${PERL_VERSION:?PERL_VERSION undefined}"
 module load bedtools/"${BEDTOOLS_VERSION:?BEDTOOLS_VERSION undefined}"
 module load pypy/"${PYPY_VERSION:?PYPY_VERSION undefined}"
 
-source /odcf/cluster/virtualenvs/warsow/python_2.7.9_SNVCalling/bin/activate
+set +ue
+source /odcf/cluster/virtualenvs/paramasi/python_2.7.9_SNVCalling_pysam_0.15.2/bin/activate
```
Comment (vinjana) [inline resources/analysisTools/snvPipeline/environments/tbi-lsf-cluster.sh]: I understand that this is the easiest solution, but maybe a more portable, explicit, and safe solution is to add the specification of the virtual env as file (similar conda env export) into the repository.

... and did you actually consider using conda?

File: resources/configurationFiles/analysisSNVCalling.xml [hunk 6]
```diff
@@ -85,13 +87,13 @@
 
             <!--## Filter criteria -->
             <!--## The rows with corresponding values larger than below criteria will be filtered out. -->
-            <cvalue name="CRIT_ExAC_maxMAF" value="0.001" type="string" />
+            <cvalue name="CRIT_ExAC_maxMAF" value="1.0" type="string" />
```
Comment (vinjana) [inline resources/configurationFiles/analysisSNVCalling.xml]: Whoa -- what a change? I guess this is tested, and not just a type or test setting.

File: resources/analysisTools/snvPipeline/environments/tbi-lsf-cluster.sh (hunk 5 above)
Comment (NagaComBio) [reply resources/analysisTools/snvPipeline/environments/tbi-lsf-cluster.sh:22]: Yes, I did try to port this to Conda when I initially started to work on this pipeline, but due to some bash issues, I faced problems running the workflow. So, I ended up copying the @GWarsow virtual environment. Maybe in a future version, this could be considered.

File: resources/configurationFiles/analysisSNVCalling.xml (hunk 6 above)
Comment (NagaComBio) [reply resources/configurationFiles/analysisSNVCalling.xml:93]: ExAC variants were included in the gnomAD WES dataset, so we just use them for annotations here, not used for the filtering.

**Comment by NagaComBio** [comment]
Thanks, @vinjana. I have tested this version of the pipeline on various projects/samples and the numbers of high confidence variants are similar to the latest version in the master. Since @GWarsow is busy, it would be better if we could find someone else to test and push the changes ASAP into the master and assign a tag.
